File: 623-add-one-row-to-tree/623-add-one-row-to-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func addOneRow(root *TreeNode, val int, dep int) *TreeNode {
    

    if dep == 1 {
        return &TreeNode{ val , root , nil }
    }
    q := []*TreeNode{root}
    for { 
        n := len(q)
        if dep ==2 {
            for i:= 0 ; i < n ; i++ { 
                pop := q[0]
                q = q[1:] 
                left , right := pop.Left , pop.Right 
                pop.Left , pop.Right = &TreeNode{ val, left , nil }, &TreeNode{ val , nil , right } 
            }
            break 
        } else { 
            for  i:= 0 ; i < n ; i++ { 
                pop := q[0] 
                q = q[1:] 
                if pop.Left != nil { q = append(q ,pop.Left ) } 
                if pop.Right != nil { q = append( q , pop.Right ) } 
            }
            dep -- 
        }
    }
    return root 

    
    
}

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.cpp

class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int dep) {
        if (dep == 0 || dep == 1) {
            TreeNode* nroot = new TreeNode(val);
            (dep ? nroot->left : nroot->right) = root;
            return nroot;
        }
        if (root && dep >= 2) {
            root->left  = addOneRow(root->left,  val, dep > 2 ? dep - 1 : 1);
            root->right = addOneRow(root->right, val, dep > 2 ? dep - 1 : 0);
        }
        return root;
    }
};